### PR TITLE
Parser Printer blank line insertion

### LINKED ISSUE #843: Apply heuristics to determine when to insert blank lines whilst printing code
Since we currently are unable to track blank lines when parsing and printing code, we should apply a series of heuristics to determine when we should insert them.

### PR DESCRIPTION
**Pull request by Rheeseyb**
Fixes #843 

**Problem:**
We print code without any blank lines between top level statements because we're unable to track their existence during parsing.

**Fix:**
Apply a series of rules to determine when to insert them

**Commit Details:**
- Updated the printing logic in `parser-printer.ts` to apply rules and optionally insert a blank line
- Added a test and fixed all existing tests that were broken by this

### CODE REVIEW

File: editor/src/core/workers/parser-printer/parser-printer.ts [hunk 1]
```diff
@@ -654,6 +655,74 @@ export const printCode = memoize(printCodeImpl, {
   maxSize: 1,
 })
 
+function isVariableDeclaration(line: string): boolean {
```
Comment (seanparsons) [inline editor/src/core/workers/parser-printer/parser-printer.ts:658]: This looks really brittle to me! It's probably fine for this use case but I can certainly imagine this is going to break on slightly different code.

File: editor/src/core/workers/parser-printer/parser-printer.ts [hunk 2]
```diff
@@ -654,6 +655,74 @@ export const printCode = memoize(printCodeImpl, {
...
+
+function isFunctionDeclaration(line: string): boolean {
+  const withoutExport = line.startsWith('export ') ? line.slice(7) : line
+  const withoutDefault = withoutExport.startsWith('default ')
+    ? withoutExport.slice(8)
+    : withoutExport
+  return withoutDefault.startsWith('function')
+}
+
+function createBlanklineTracker() {
```
Comment (seanparsons) [inline editor/src/core/workers/parser-printer/parser-printer.ts:675]: Would it be possible to have a return type on this function please?

File: editor/src/core/workers/parser-printer/parser-printer.ts (hunk 1 above)
Comment (Rheeseyb) [reply editor/src/core/workers/parser-printer/parser-printer.ts:658]: Yeah, I think this whole PR is one of those "it's not great but better than nothing" situations unfortunately :grimacing:

File: editor/src/core/workers/parser-printer/parser-printer.ts (hunk 2 above)
Comment (Rheeseyb) [reply editor/src/core/workers/parser-printer/parser-printer.ts]: Done

File: editor/src/core/workers/parser-printer/parser-printer.ts (hunk 1 above)
Comment (seanparsons) [reply editor/src/core/workers/parser-printer/parser-printer.ts:658]: That's cool.

**Review by seanparsons — APPROVED** [review]

**Review by balazsbajorics — COMMENTED** [review]
:+1:

**Pull request merged**